mkcodes.py

```diff
@@ -13,48 +13,79 @@
     from markdown.extensions import Extension
     from markdown.treeprocessors import Treeprocessor
 
-
-def github_codeblocks(filepath, safe):
-    codeblocks = []
+# much easier to write the other names that an extension is known by
+ext_map = {
+    'cs': ['c#', 'csharp', 'c-sharp', 'cs', ],
+    'java': ['java',],
+    'py': ['python', 'py', 'python2', 'python3', 'py2', 'py3',
+           'Python'],
+}
+# then invert that mapping
+language_map = {}
+for ext, lang_strings in ext_map.items():
+    for lang_string in lang_strings:
+        language_map[lang_string] = ext
+
+
+def github_codeblocks(filepath, safe, default_lang='py'):
+    codeblocks = {}
     codeblock_re = r'^```.*'
-    codeblock_open_re = r'^```(`*)(py|python){0}$'.format('' if safe else '?')
+    codeblock_open_re = r'^```(`*)(\w+){0}$'.format('' if safe else '?')
 
     with open(filepath, 'r') as f:
         block = []
-        python = True
+        language = None
         in_codeblock = False
 
         for line in f.readlines():
+            # does this line contain a codeblock begin or end?
             codeblock_delimiter = re.match(codeblock_re, line)
 
             if in_codeblock:
                 if codeblock_delimiter:
-                    if python:
-                        codeblocks.append(''.join(block))
+                    # we are closing a codeblock
+                    if language:
+                        # finished a codeblock, append everything
+                        ext = language_map.get(language, language)
+                        codeblocks.setdefault(ext, []).append(''.join(block))
+
                     block = []
-                    python = True
+                    if safe:
+                        language = None
                     in_codeblock = False
                 else:
                     block.append(line)
             elif codeblock_delimiter:
+                # beginning a codeblock
                 in_codeblock = True
-                if not re.match(codeblock_open_re, line):
-                    python = False
+                # does it have a language?
+                lang_match = re.match(codeblock_open_re, line)
+                if lang_match:
+                    language = lang_match.group(2).lower()
+                    if not safe:
+                        # we can sub a default language if not safe
+                        language = language or default_lang
+                else:
```

Comment:
It seems that this entire `else` block can be deleted as no-op. if there's no language specified then `language = None` is fine -- `language = language or default_lang` will pick up the `default_lang` if the safety is off.